pym/bob/input.py

```diff
@@ -576,6 +577,62 @@ def isDeterministic(self):
     def hasJenkinsPlugin(self):
         return True
 
+    def callGit(self, workspacePath, *args):
+        cmdLine = ['git']
+        for arg in args:
+            cmdLine.append(arg)
+        try:
+            output = subprocess.check_output(cmdLine, cwd=os.path.join(os.getcwd(),workspacePath, self.__dir),
+                universal_newlines=True, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            raise BuildError("git error: '{}' '{}'".format(" ".join(cmdLine), e.output))
+        return output
+
+    def status(self, workspacePath, dir, verbose):
+        if not os.path.exists(os.path.join(os.getcwd(), workspacePath, self.__dir)):
+            return 'error'
+
+        ret = 0
+        if self.__commit:
+            output = self.callGit(workspacePath, 'rev-parse', 'HEAD').rstrip()
+            if output != self.__commit:
+                ret += 1
+                if verbose:
+                    print(colorize("    ! should have commitId {} but commitID is {}".format(self.__commit, output), "33"))
+        elif self.__tag:
+            output = self.callGit(workspacePath, 'describe', '--tags', '--always').rstrip()
+            if output != self.__tag:
+                ret += 1
+                if verbose:
+                    print(colorize("    ! should be on tag {} but tag is {}".format(self.__tag, output), "33"))
+        elif self.__branch:
+            output = self.callGit(workspacePath, 'rev-parse', '--abbrev-ref', 'HEAD').rstrip()
+            if output != self.__branch:
+                ret += 1
+                if verbose:
+                    print(colorize("    ! should be on branch {} but currently on {}".format(self.__branch, output), "33"))
+
+        output = self.callGit(workspacePath, 'status', '--porcelain')
+        if len(output):
+            if verbose:
+                print(colorize("    ! has modifications:", "33"))
+                for line in output.split('\n'):
+                    print('      '+line.rstrip())
+            ret += 1
+
+        output = self.callGit(workspacePath, 'log', '--all', '--not', '--remotes')
+        if len(output):
+            if verbose:
+                print(colorize("    ! has unpushed commit(s):", "33"))
+                for line in output.split('\n'):
+                    print('      '+line.rstrip())
+            ret += 1
```

Comment:
Depends on how you define the clean-checkout option I think. This check shows commits you haven't pushed to upstream and you build from other sources then the rest of your team, CI server or whatever. IMO this should be resolved by the clean-checkout option.